core/src/sumeragi/main_loop.rs

```diff
@@ -448,60 +413,71 @@ fn replace_top_block<F: FaultInjection>(
     block: impl Into<VersionedCommittedBlock>,
 ) {
     let committed_block = block.into();
-    let block_hash = committed_block.hash();
 
     state.wsv = state.finalized_wsv.clone();
-    state
-        .wsv
-        .apply(&committed_block)
-        .expect("Failed to apply block on WSV. Bailing.");
-
-    sumeragi.send_events(state.wsv.events_buffer.replace(Vec::new()));
-
-    // Update WSV copy that is public facing
-    *sumeragi.wsv.lock() = state.wsv.clone();
-
-    // This sends "Block committed" event, so it should be done
-    // AFTER public facing WSV update
-    sumeragi.send_events(&committed_block);
-
+    update_state(state, sumeragi, &committed_block);
     // state.previous_block_hash stays the same.
-    state.latest_block_height = committed_block.header().height;
-    state.latest_block_hash = Some(block_hash);
-    state.latest_block_view_change_index = committed_block.header().view_change_index;
-
-    let current_topology = &mut state.current_topology;
-    let role = current_topology.role(&sumeragi.peer_id);
 
     info!(
-        addr=%sumeragi.peer_id.address, %role,
+        addr=%sumeragi.peer_id.address,
+        role=%state.current_topology.role(&sumeragi.peer_id),
         block_height=%state.latest_block_height,
-        %block_hash, "Replacing top block"
+        block_hash=%committed_block.hash(),
+        "Replacing top block"
     );
 
-    *current_topology = Topology {
+    update_topology(state, &committed_block);
+
+    sumeragi.kura.replace_top_block(committed_block);
+
+    cache_transaction(state, sumeragi)
+}
+
+fn update_topology(state: &mut State, committed_block: &VersionedCommittedBlock) {
```

Comment:
Seems it should be a method of `State`, but I suppose you have a strong reason to keep it as a separate function, right?

Reply:
I agree that this is a mutator method and it **should** be associated to state (i.e. written in an `impl` block). 

> other functions that also looks like "method" functions, but some reason we keep them as separate functions.

This is mostly stylistic, but I can give you a rational reason why that's a good idea. Functions associated to types are in a privileged position with respect to access modifiers. That's why you need to be careful about the visibility of these functions. Everything is a method is an OOP concept, and not necessarily a useful one. It can help readability, and it definitely would in this particular case, but sometimes it's not very helpful. 

IMO, given that we're refactoring the file, we should try and look at other potentially associated `fn`s and try to add them to the respective `impl` blocks.